compiler/vm.nim

```diff
@@ -746,10 +746,13 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
       regs[ra].regAddr = addr(regs[rb])
     of opcAddrNode:
       decodeB(rkNodeAddr)
-      if regs[rb].kind == rkNode:
+      case regs[rb].kind
+      of rkNode:
         regs[ra].nodeAddr = addr(regs[rb].node)
+      of rkNodeAddr: # bug #14339
+        regs[ra].nodeAddr = regs[rb].nodeAddr
```

Comment:
Doesn't that mean that when you do `somePtr = addr someOtherPtr` it will behave like `somePtr = someOtherPtr`?

Reply:
I don't believe so; eg:
```nim
when true:
  type Foo = ref object
    f0: int
  proc main=
    var f = Foo(f0: 3)
    var f2 = f.addr
    var f3 = f2.addr
    var f4 = f3
    f4.f0 += 1 # like f3.f0
    f4[].f0 += 1 # like f2.f0
    f4[][].f0 += 1 # like f.f0
    f4[][][].f0 += 1 # like f[].f0
    doAssert f[].f0 == 7
  static: main()
  main()
```

=> CT behaves like RT

Reply:
Hmm, I'm struggling to come up with a concrete example, but please explain why it is correct and maybe add a comment.
Your example isn't exactly convincing because `f4` is `ptr ptr Foo` and it gets automatically dereferenced. If you change `Foo` to be a non-ref object it fails.

Reply:
> If you change Foo to be a non-ref object it fails.

just filed https://github.com/nim-lang/Nim/issues/16020, but this is a pre-existing bug and this PR exercises a different code path, see https://gist.github.com/timotheecour/fc575c054cddba5bea32df98d1c50cdc

> Hmm, I'm struggling to come up with a concrete example, but please explain why it is correct (?) and maybe add a comment.

TODO on me

Reply:
@Clyybber my understanding is there's an implicit deref that occurs in code paths that lead here due to semantic transform that generates a temporary for block return expressions (without `result = expr` or return expr`)
eg:

```nim
when true: # D20201125T111240
  template main() =
    type
      Node = ref object
        val: int
    proc bar(c: Node): var int =
      discard
      c.val #  before PR: Error: limited VM support for 'addr' (but nim was pointing at wrong lineinfo)
    var a = Node()
    discard a.bar()
  static: main()
```

this seems related to https://github.com/nim-lang/Nim/issues/16137

This PR is still good IMO (more things now work, and hopefully nothing breaks), but I agree that the code transformation needs to be revisited as @cooldome mentioned here: https://github.com/nim-lang/Nim/issues/14420#issuecomment-633711557

Reply:
The code transformation is correct, and is not what my question is about.
I'm asking how treating `... = addr someAddress` as if it were `... = someAddress` is correct/sound.